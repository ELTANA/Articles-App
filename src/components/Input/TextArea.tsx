import { FC, forwardRef } from 'react';
import type { TextAreaProps } from './Input.types';
import useInputError from '$hooks/useInputError';

const TextArea: FC<TextAreaProps> = forwardRef<HTMLDivElement, TextAreaProps>(
  ({ error, label, name, textAreaRef, ...rest }, ref) => {
    const { errorComponent } = useInputError({ error });
    return (
      <div ref={ref} className="flex flex-col gap-4 w-full">
        {label ? (
          <label className="text-sm md:text-base text-gray-600 capitalize font-medium" htmlFor={name}>
            {label}
          </label>
        ) : null}
        <div className="w-full relative h-max">
          <textarea
            ref={textAreaRef}
            name={name}
            id={name}
            className={`w-full shadow-sm bg-white rounded border text-base outline-none text-gray-700 py-1 px-3 leading-8 transition-colors duration-200 ease-in-out resize-none h-[150px] ${error ? 'border-red-200 focus:border-red-300 focus:ring-2 focus:ring-red-300' : 'border-blue-200 focus:border-blue-100 focus:ring-2 focus:ring-blue-100'}`}
            aria-label={label}
            {...rest}
          ></textarea>
        </div>
        {errorComponent}
      </div>
    );
  }
);

TextArea.displayName = 'TextArea';

export default TextArea;
