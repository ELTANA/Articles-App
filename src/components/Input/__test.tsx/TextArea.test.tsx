import { render, screen, waitFor } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import TextArea from '../TextArea';
import '@testing-library/jest-dom';

beforeAll(() => {
  window.ResizeObserver =
    window.ResizeObserver ||
    jest.fn().mockImplementation(() => ({
      disconnect: jest.fn(),
      observe: jest.fn(),
      unobserve: jest.fn()
    }));
});

describe('TextArea component', () => {
  test('renders TextArea with label and without error', () => {
    render(<TextArea label="Test Label" name="testTextArea" />);

    const labelElement = screen.getByText(/Test Label/i);
    const textAreaElement = screen.getByRole('textbox', { name: 'Test Label' });

    expect(labelElement).toBeInTheDocument();
    expect(textAreaElement).toBeInTheDocument();
  });

  test('renders TextArea with error', () => {
    render(<TextArea label="Test Label" name="testTextArea" error={{ type: 'value', message: 'Error happened' }} />);

    const errorElement = screen.getByText(/Error happened/i);

    expect(errorElement).toBeInTheDocument();
  });

  test('handles user input', async () => {
    const handleChange = jest.fn();
    render(<TextArea label="Test Label" name="testTextArea" onChange={handleChange} />);

    const textAreaElement = screen.getByRole('textbox', { name: 'Test Label' });

    userEvent.type(textAreaElement, 'Hello');

    // Use waitFor to wait for asynchronous changes
    await waitFor(() => {
      // Expect handleChange to be called for each character in 'Hello'
      expect(handleChange).toHaveBeenCalledTimes(5); // Each keystroke triggers onChange
    });
    expect(textAreaElement).toHaveValue('Hello');
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<TextArea label="Test Label" name="testTextArea" />);

    expect(asFragment()).toMatchSnapshot();
  });
});
