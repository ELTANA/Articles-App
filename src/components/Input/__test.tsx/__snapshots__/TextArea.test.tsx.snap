// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4 w-full"
  >
    <label
      class="text-sm md:text-base text-gray-600 capitalize font-medium"
      for="testTextArea"
    >
      Test Label
    </label>
    <div
      class="w-full relative h-max"
    >
      <textarea
        aria-label="Test Label"
        class="w-full shadow-sm bg-white rounded border text-base outline-none text-gray-700 py-1 px-3 leading-8 transition-colors duration-200 ease-in-out resize-none h-[150px] border-blue-200 focus:border-blue-100 focus:ring-2 focus:ring-blue-100"
        id="testTextArea"
        name="testTextArea"
      />
    </div>
    <div
      class="h-fit transition-all"
    >
      <p
        class="font-normal text-sm leading-4 text-red-600 transition-all"
      />
    </div>
  </div>
</DocumentFragment>
`;
